### PR TITLE
[package] PackageExporter visualization methods

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61148
* **#61147**
* #61146

Basic tooling to enable users to see what is inside of a PackageExporter. Added methods: 
- `externed/interned/mocked/denied_list()`: returns list of modules which are currently in the specified category
- `relied_on_by(module_name)`: returns list of modules which rely on `module_name`
- `dependency_graph_str()`: returns string format of graph for users. Example of output:
```
digraph G {
rankdir = LR;
node [shape=box];
"<res.foo.pkl>" -> "foo";
"foo" -> "torch.package";
"foo" -> "time";
"foo" -> "sentencepiece";
"foo" -> "package_top";
}
```

Differential Revision: [D29559683](https://our.internmc.facebook.com/intern/diff/D29559683)

### CODE REVIEW

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559683).

File: torch/package/package_exporter.py [hunk 1]
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
...
+    ) -> List[str]:
+        result = []
+        for name, node_dict in self.dependency_graph.nodes.items():
+            node_action = node_dict.get("action", None)
+            if node_action == action and "is_pickle" not in node_dict:
+                result.append(name)
+        result.sort()
+        return result
+
+    def externed_list(self) -> List[str]:
```
Comment (suo) [inline torch/package/package_exporter.py:926]: maybe `externed_modules` and so on?

File: torch/package/package_exporter.py [hunk 2]
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
...
+    def denied_list(self) -> List[str]:
+        """Return all modules that are currently denied.
+
+        Returns:
+            A list containing the names of modules which will be
+            denied in this package.
+        """
+        return self._nodes_with_action_type(_ModuleProviderAction.DENY)
+
+    def relied_on_by(self, module_name: str) -> List[str]:
```
Comment (suo) [inline torch/package/package_exporter.py:962]: The "rely" language is not previously present in the API/docs—I think we should try to stay consistent with talking about "dependencies". The typical term is "reverse dependency", so maybe: `get_reverse_dependencies` or `get_rdeps`,

File: torch/package/package_exporter.py [hunk 3]
```diff
@@ -902,6 +902,74 @@ def _can_implicitly_extern(self, module_name: str):
...
+        """Return a list of all modules which rely on the module ``module_name``.
```
Comment (suo) [inline torch/package/package_exporter.py:963]: rely -> depend

**Review by suo — CHANGES_REQUESTED** [review]
Overall looks good to me, had some small comments to improve API consistency

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559683).

**Review by suo — APPROVED** [review]

**Pull request closed without merging**